core/templates/dev/head/pages/exploration_player/CurrentInteractionService.js

```diff
@@ -0,0 +1,100 @@
+ * while the latter contains the actual "Submit" button which triggers the
+ * answer submission process.
+ */
+
+oppia.factory('CurrentInteractionService', [
+  function() {
+    var _submitAnswer;
+    var _onSubmit;
+    var _validityCheckFn;
+    var _preSubmitHooks = [];
+
+    return {
+      setOnSubmitFn: function(onSubmit) {
+        /**
+         * The ConversationSkinDirective should register its onSubmit
+         * callback here.
+         *
+         * @param {function(answer, interactionRulesService)} onSubmit
+         */
+        _onSubmit = onSubmit;
+      },
+      registerCurrentInteraction: function(submitAnswer, validityCheckFn) {
+        /**
+         * Every interaction directive should register themselves here,
+         * even those which do not use the progress nav. All arguments are
+         * optional.
+         *
+         * @param {function} submitAnswer - Should grab the learner's answer
+         *   and pass it to onSubmit. The interaction can omit this param if
+         *   it does not use the progress nav's submit button
+         *   (ex: MultipleChoiceInput).
+         * @param {function} validityCheckFn - The progress nav will use this
+         *   to decide whether or not to disable the submit button. The
+         *   interaction can omit this param if it does not use the progress
+         *   nav's submit button.
+         */
+        _submitAnswer = submitAnswer;
+        _validityCheckFn = validityCheckFn;
+      },
+      registerPreSubmitHook: function(hookFn) {
+        /* Register a time hook that will be called right before onSubmit.
+         * All hooks for the current interaction will be cleared right
+         * before loading the next card.
+         */
+        _preSubmitHooks.push(hookFn);
+      },
+      clearPreSubmitHooks: function() {
+        /* Clear out all the hooks for the current interaction. Should
+         * be called before loading the next card.
+         */
+        _preSubmitHooks = [];
+      },
+      onSubmit: function(answer, interactionRulesService) {
+        for (var i = 0; i < _preSubmitHooks.length; i++) {
+          hookFn = _preSubmitHooks[i];
+          hookFn();
+        }
+        _onSubmit(answer, interactionRulesService);
+      },
+      submitAnswer: function() {
+        /* This starts the answer submit process, it should be called once the
+         * learner presses the "Submit" button.
+         */
+        if (typeof _submitAnswer === 'undefined') {
+          throw Error('The current interaction did not ' +
+                      'register a submitAnswer function.');
+        } else {
+          _submitAnswer();
+        }
+      },
+      isSubmitButtonDisabled: function() {
+        /* Returns whether or not the Submit button should be disabled based on
+         * the validity of the current answer.
+         */
+        if (typeof _validityCheckFn === 'undefined') {
+          return false;
```

Comment:
Why not throw error?

Reply:
Not every interaction passes in a validityCheckFn, so if one doesn't exist all answers are assumed to be valid.

Alternatively, we could require every interaction have a validityCheckFn and for all the interactions that don't have one yet just define a dummy that always returns true.

Reply:
Let's do the latter -- more consistent, I think.

Reply:
Having every interaction define a dummy validity check felt repetitive, so instead I've created a default validity function in CurrentInteractionService. Also, interactions which don't have a validity check now pass null in explicitly.